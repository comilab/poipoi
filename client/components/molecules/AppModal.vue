<template lang="pug">
transition(
  enter-class="opacity-0",
  leave-to-class="opacity-0"
)
  .fixed.inset-0.bg-black.bg-opacity-75.z-50.translate-opacity.duration-300(
    v-if="innerValue",
    @click.self="!persistent && close()",
    data-testid="overlay"
  )
    .position-center.bg-white.rounded.max-w-screen-md.max-h-90.p-4
      AppButton.absolute.top-0.right-0.-mx-4.-my-4.shadow(
        v-if="!persistent",
        color="blue",
        size="sm",
        :icon="faTimes",
        pill,
        @click="close",
        aria-label="close",
        data-testid="close-button"
      )
      .overflow-auto(data-testid="content")
        slot(:close="close")
</template>

<script lang="ts">
import { defineComponent } from 'nuxt-composition-api'
import { faTimes } from '@fortawesome/free-solid-svg-icons'

import useVModel from '~/composables/use-v-model'
import AppButton from '~/components/atoms/AppButton.vue'

export default defineComponent({
  components: {
    AppButton
  },
  props: {
    value: {
      type: Boolean,
      required: false,
      default: false
    },
    persistent: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  setup (props, { emit }) {
    const { innerValue } = useVModel(props, emit)

    const open = () => {
      innerValue.value = true
      emit('opened')
    }

    const close = () => {
      innerValue.value = false
      emit('closed')
    }

    return {
      innerValue,
      open,
      close,
      faTimes
    }
  }
})
</script>
