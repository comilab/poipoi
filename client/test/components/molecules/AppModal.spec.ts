import { fireEvent, render, within } from '@testing-library/vue'
import { defineComponent, ref } from 'nuxt-composition-api'
import { Ref } from '@vue/composition-api'

import AppModal from '~/components/molecules/AppModal.vue'

describe('components/molecules/AppModal', () => {
  let options: any

  beforeEach(() => {
    options = {
      props: {
        value: true
      },
      stubs: {
        FontAwesomeIcon: true
      },
      scopedSlots: {}
    }
  })

  test('renders correctly', () => {
    const { container } = render(AppModal, options)

    expect(container).toMatchSnapshot()
  })

  describe('props', () => {
    describe('value', () => {
      test('trueならモーダルを表示', () => {
        options.props.value = true
        const { container, queryByTestId } = render(AppModal, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('overlay')).toBeInTheDocument()
      })

      test('falseならモーダルを非表示', () => {
        options.props.value = false
        const { container, queryByTestId } = render(AppModal, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('overlay')).not.toBeInTheDocument()
      })

      test('デフォルトはfalse', () => {
        options.props = {}
        const { queryByTestId } = render(AppModal, options)

        expect(queryByTestId('overlay')).not.toBeInTheDocument()
      })
    })

    describe('persistent', () => {
      test('trueなら閉じるボタンを表示しない', () => {
        options.props.persistent = true
        const { container, queryByTestId } = render(AppModal, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('close-button')).not.toBeInTheDocument()
      })

      test('trueならオーバーレイをクリックしても閉じない', async () => {
        options.props.persistent = true
        const { getByTestId, emitted } = render(AppModal, options)
        await fireEvent.click(getByTestId('overlay'))

        expect(emitted().closed).toBeFalsy()
      })

      test('falseなら閉じるボタンを表示', () => {
        options.props.persistent = false
        const { container, queryByTestId } = render(AppModal, options)

        expect(container).toMatchSnapshot()
        expect(queryByTestId('close-button')).toBeInTheDocument()
      })

      test('デフォルトはfalse', () => {
        const { queryByTestId } = render(AppModal, options)

        expect(queryByTestId('close-button')).toBeInTheDocument()
      })
    })
  })

  describe('slots', () => {
    describe('default', () => {
      test('content内に表示', () => {
        options.scopedSlots = {
          default: '<div>slot-default</div>'
        }
        const { getByTestId } = render(AppModal, options)

        within(getByTestId('content')).getByText('slot-default')
      })

      test('scope.closeを実行するとモーダルを閉じる', async () => {
        options.scopedSlots = {
          default: `
            <div slot-scope="{ close }">
              <button data-testid="slot-close-button" @click="close" />
            </div>
          `
        }
        const { getByTestId, emitted } = render(AppModal, options)
        expect(getByTestId('slot-close-button')).toBeInTheDocument()

        await fireEvent.click(getByTestId('slot-close-button'))

        expect(emitted().closed).toHaveLength(1)
      })
    })
  })

  describe('open', () => {
    let wrapper: any
    let appModal: Ref<any>
    let isOpen: Ref<boolean>
    let onInput: jest.Mock
    let onOpened: jest.Mock

    beforeEach(() => {
      onInput = jest.fn()
      onOpened = jest.fn()
      wrapper = defineComponent({
        components: { AppModal },
        setup () {
          appModal = ref()
          isOpen = ref(false)
          return { appModal, isOpen, onInput, onOpened }
        },
        template: `
          <div>
            <AppModal
              ref="appModal"
              v-model="isOpen"
              @input="onInput"
              @opened="onOpened"
            />
          </div>
        `
      })
    })

    test('実行するとvalue=trueになる', () => {
      render(wrapper, options)
      appModal.value.open()

      expect(isOpen.value).toBeTruthy()
    })

    test('inputイベントを発火', () => {
      render(wrapper, options)
      appModal.value.open()

      expect(onInput).toBeCalledWith(true)
    })

    test('openedイベントを発火', () => {
      render(wrapper, options)
      appModal.value.open()

      expect(onOpened).toBeCalled()
    })
  })

  describe('close', () => {
    let wrapper: any
    let appModal: Ref<any>
    let isOpen: Ref<boolean>
    let onInput: jest.Mock
    let onClosed: jest.Mock

    beforeEach(() => {
      onInput = jest.fn()
      onClosed = jest.fn()
      wrapper = defineComponent({
        components: { AppModal },
        setup () {
          appModal = ref()
          isOpen = ref(true)
          return { appModal, isOpen, onInput, onClosed }
        },
        template: `
          <div>
            <AppModal
              ref="appModal"
              v-model="isOpen"
              @input="onInput"
              @closed="onClosed"
            />
          </div>
        `
      })
    })

    test('実行するとvalue=falseになる', () => {
      render(wrapper, options)
      appModal.value.close()

      expect(isOpen.value).toBeFalsy()
    })

    test('inputイベントを発火', () => {
      render(wrapper, options)
      appModal.value.close()

      expect(onInput).toBeCalledWith(false)
    })

    test('closedイベントを発火', () => {
      render(wrapper, options)
      appModal.value.close()

      expect(onClosed).toBeCalled()
    })
  })

  describe('event', () => {
    test('オーバーレイをクリックすると閉じる', async () => {
      const { getByTestId, emitted } = render(AppModal, options)
      await fireEvent.click(getByTestId('overlay'))

      expect(emitted().closed).toHaveLength(1)
    })

    test('閉じるボタンをクリックすると閉じる', async () => {
      const { getByTestId, emitted } = render(AppModal, options)
      await fireEvent.click(getByTestId('close-button'))

      expect(emitted().closed).toHaveLength(1)
    })
  })
})
