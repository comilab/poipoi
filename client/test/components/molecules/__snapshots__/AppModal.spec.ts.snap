// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/AppModal props persistent falseなら閉じるボタンを表示 1`] = `
<div>
  <transition-stub
    enterclass="opacity-0"
    leavetoclass="opacity-0"
  >
    <div
      class="fixed inset-0 bg-black bg-opacity-75 z-50 translate-opacity duration-300"
      data-testid="overlay"
    >
      <div
        class="position-center bg-white rounded max-w-screen-md max-h-90 p-4"
      >
        <button
          aria-label="close"
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none absolute top-0 right-0 -mx-4 -my-4 shadow bg-blue-600 text-white hover_bg-blue-700 text-sm px-2 py-1 square-sm rounded-full"
          data-testid="close-button"
          type="button"
        >
          <div>
            <fontawesomeicon-stub
              class="fa-fw"
              data-testid="icon"
              icon="[object Object]"
            />
          </div>
          <div
            class=""
            data-testid="text"
          />
        </button>
        <div
          class="overflow-auto"
          data-testid="content"
        />
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`components/molecules/AppModal props persistent trueなら閉じるボタンを表示しない 1`] = `
<div>
  <transition-stub
    enterclass="opacity-0"
    leavetoclass="opacity-0"
  >
    <div
      class="fixed inset-0 bg-black bg-opacity-75 z-50 translate-opacity duration-300"
      data-testid="overlay"
    >
      <div
        class="position-center bg-white rounded max-w-screen-md max-h-90 p-4"
      >
        <!---->
        <div
          class="overflow-auto"
          data-testid="content"
        />
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`components/molecules/AppModal props value falseならモーダルを非表示 1`] = `
<div>
  <transition-stub
    enterclass="opacity-0"
    leavetoclass="opacity-0"
  >
    <!---->
  </transition-stub>
</div>
`;

exports[`components/molecules/AppModal props value trueならモーダルを表示 1`] = `
<div>
  <transition-stub
    enterclass="opacity-0"
    leavetoclass="opacity-0"
  >
    <div
      class="fixed inset-0 bg-black bg-opacity-75 z-50 translate-opacity duration-300"
      data-testid="overlay"
    >
      <div
        class="position-center bg-white rounded max-w-screen-md max-h-90 p-4"
      >
        <button
          aria-label="close"
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none absolute top-0 right-0 -mx-4 -my-4 shadow bg-blue-600 text-white hover_bg-blue-700 text-sm px-2 py-1 square-sm rounded-full"
          data-testid="close-button"
          type="button"
        >
          <div>
            <fontawesomeicon-stub
              class="fa-fw"
              data-testid="icon"
              icon="[object Object]"
            />
          </div>
          <div
            class=""
            data-testid="text"
          />
        </button>
        <div
          class="overflow-auto"
          data-testid="content"
        />
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`components/molecules/AppModal renders correctly 1`] = `
<div>
  <transition-stub
    enterclass="opacity-0"
    leavetoclass="opacity-0"
  >
    <div
      class="fixed inset-0 bg-black bg-opacity-75 z-50 translate-opacity duration-300"
      data-testid="overlay"
    >
      <div
        class="position-center bg-white rounded max-w-screen-md max-h-90 p-4"
      >
        <button
          aria-label="close"
          class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none absolute top-0 right-0 -mx-4 -my-4 shadow bg-blue-600 text-white hover_bg-blue-700 text-sm px-2 py-1 square-sm rounded-full"
          data-testid="close-button"
          type="button"
        >
          <div>
            <fontawesomeicon-stub
              class="fa-fw"
              data-testid="icon"
              icon="[object Object]"
            />
          </div>
          <div
            class=""
            data-testid="text"
          />
        </button>
        <div
          class="overflow-auto"
          data-testid="content"
        />
      </div>
    </div>
  </transition-stub>
</div>
`;
